sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -519,7 +519,9 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
 
   override def sql: String = {
     val valueSQL = child.sql
-    val listSQL = hset.toSeq.map(Literal(_).sql).mkString(", ")
```

Comment:
what's wrong with `Literal.sql`?

Reply:
It doesn't accept UTF8String at https://github.com/apache/spark/blob/ff39c9271ca04951b045c5d9fca2128a82d50b46/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L292

Reply:
This seems to be an orthogonal fix. @cloud-fan and @MaxGekk . We need this in `branch-2.4` because this is SPARK-12593 (Converts resolved logical plan back to SQL) since Apache Spark 2.0.0, don't we?

If you don't mind, please file a separate JIRA issue with a separate test case. We need to merge this seperately.
cc @holdenk

Reply:
Please let me know the JIRA when you file it and I'll add it to my tracking for the 2.4.6 release.

Reply:
> If you don't mind, please file a separate JIRA issue with a separate test case. We need to merge this seperately.

So far, I have not found how to trigger the issue in the `sql` method without this fix. I will think of that and try tomorrow but if you have some ideas you are welcome.

What I have already tried is to build a dataset with IsIn, optimizer converts it to InSet but I wasn't able to call `sql()` on the replaced expression.

Reply:
Here is JIRA ticket https://issues.apache.org/jira/browse/SPARK-31563

Reply:
Here is the PR https://github.com/apache/spark/pull/28343